object_storage_api/repositories/image.py

```diff
@@ -42,20 +43,27 @@ def create(self, image: ImageIn, session: ClientSession = None) -> ImageOut:
         result = self._images_collection.insert_one(image.model_dump(by_alias=True), session=session)
         return self.get(str(result.inserted_id), session=session)
 
-    def get(self, image_id: str, session: ClientSession = None) -> Optional[ImageOut]:
+    def get(self, image_id: str, session: ClientSession = None) -> ImageOut:
         """
         Retrieve an image by its ID from a MongoDB database.
 
         :param image_id: ID of the image to retrieve.
         :param session: PyMongo ClientSession to use for database operations.
-        :return: Retrieved image or `None` if not found.
+        :return: Retrieved image if found.
+        :raises MissingRecordError: If the supplied `image_id` is non-existent.
+        :raises InvalidObjectIdError: If the supplied `image_id` is invalid.
         """
-        image_id = CustomObjectId(image_id)
         logger.info("Retrieving image with ID: %s from the database", image_id)
-        image = self._images_collection.find_one({"_id": image_id}, session=session)
+        try:
+            image_id = CustomObjectId(image_id)
+            image = self._images_collection.find_one({"_id": image_id}, session=session)
+        except InvalidObjectIdError as exc:
+            exc.status_code = 404
+            exc.response_detail = "Image not found"
```

Comment:
What are the cases where `InvalidObjectIdError` has a different `status_code` and `response_detail`?

Reply:
@VKTB The issue here is that the `InvalidObjectIdError` is raised when any schema fails to validate due to an invalid ID, normally when this happens we expect it to be a 422 as it happens in request bodies, which is why its the default when raised in `CustomObjectId`. Here we are catching it just because its a specific case for the get request where we are trying to make it return a 404 as its actually an endpoint path parameter instead. So I think its worse having it as  404, as all other requests would have to put it back to a 422 again vs one get having to do it.

Reply:
I meant something like
```python
def check_path_id(v: str) -> str:
    try
        CustomObjectId(v)
    except InvalidIdError as exc
         exc.status_code = 404
         raise exc

PathIdStr = Annotated[str, AfterValidator[check_path_id]]
```
Then all router endpoints that has an id in its path param use `PathIdStr` instead of str, so only those will validate the id in this way and the rest presumably happening in schemas are 422.

Reply:
I see what you meant. That could probably work, yes but I will leave it to you guys to decide if you want to change it especially if the current approach already works.

Reply:
Yeah it would need an entity name as well anyway so the responses matched, so it could also be preferable to keep it as is so they are both in the same place. I would probably just leave it for this PR.

Reply:
I've just pushed a commit where the validator accepts the entity name as well, at the router level.

Can roll it back to being raised at the repo level though, if we want to leave it for a separate issue.

Reply:
Yeah I would prefer it to be addressed separately, I think there could be a few ways to consider e.g. https://docs.pydantic.dev/latest/concepts/validation_decorator/#extension-validating-arguments-before-calling-a-function and we need this PR merged to unblock reviewing some front end ones. I will add it to https://github.com/ral-facilities/object-storage-api/issues/15.